### PR TITLE
Reduce indexes creation time by walking the DOM

### PR DESCRIPTION
**Pull request by frangarcj**
On Friday, the 29th of September, the registry XML size was about 16 MB, and it took like 20 minutes to process the entire catalog.

The following PR aims to reduce index creation time by walking the dom and removing some XPATH evaluations.

Execution time with current catalog (approx. 5MB):

- Original code: 30189 ms
- New code: 970 ms

### CODE REVIEW

**Comment by janinamincer-daszkiewicz** [comment]
Thank you very much for this post.
Yes, the size of the cataloque went up. This is now back to the previous size but at some moment it will become large again. We - nodes in the network - need some time to adapt. I wanted to start the GitHub issue myself but you have been faster :)  
Let's gather here recommendations for developers reading/parsing the catalogue in various languages.

File: src/main/java/eu/erasmuswithoutpaper/registryclient/CatalogueDocument.java [hunk 1]
```diff
@@ -305,14 +353,20 @@ public Iterator<String> getPrefixes(String namespaceUri) {
 
         mapForType.put(getCanonicalId(value), heiId);
       }
-      for (Element heiElem : Utils.asElementList(
```
Comment (mkurzydlowski) [inline src/main/java/eu/erasmuswithoutpaper/registryclient/CatalogueDocument.java:308]: Why did you extract variables from the following "for" loops?

Comment (frangarcj) [reply src/main/java/eu/erasmuswithoutpaper/registryclient/CatalogueDocument.java:308]: It looks strange but the debugger was evaluating the full expression for each loop

**Comment by m-sokolowski** [comment]
> Let's gather here recommendations for developers reading/parsing the catalogue in various languages.

In our Python application the time required to parse the catalogue didn't increase significantly. On my local machine it takes about 114 ms to parse the current 5MB catalogue, and 338 ms to parse that big catalogue. We use lxml:

```
from lxml import etree

doc = etree.fromstring(xml, etree.XMLParser(huge_tree=True))
```

We just had to add the `huge_tree=True` option to enable parsing such large files, because an error was thrown due to default security limits.

We don't create indexes like in ewp-registry-client. Nevertheless the simple XPath queries that we use are still running fast.

File: src/main/java/eu/erasmuswithoutpaper/registryclient/CatalogueDocument.java (hunk 1 above)
Comment (mkurzydlowski) [reply src/main/java/eu/erasmuswithoutpaper/registryclient/CatalogueDocument.java:308]: So this is only done for debugging purposes and doesn't speed up production code?

I wanted to suggest reverting those changes, as they make the code "less clean". What do you think?

**Comment by mkurzydlowski** [comment]
I have saved the big catalogue file generated after changes to Dashboard manifest, for tests:

https://schowek.usos.edu.pl/m.kurzydlowski/catalogue-v1.xml

File: src/main/java/eu/erasmuswithoutpaper/registryclient/CatalogueDocument.java (hunk 1 above)
Comment (frangarcj) [reply src/main/java/eu/erasmuswithoutpaper/registryclient/CatalogueDocument.java:308]: It is not only for debugging but also to have it cleaner.
The less you do in a line of code, the cleaner it gets.

Comment (mkurzydlowski) [reply src/main/java/eu/erasmuswithoutpaper/registryclient/CatalogueDocument.java:308]: It seems to me that it's better to not define variables that are not used outside the loop but maybe let's not dwell on that. I can accept the current change after I make sufficient tests.

Thanks for you help!

**Comment by janinamincer-daszkiewicz** [comment]
I hope the other providers will also test their solutions on this large catalogue file. Please, post here test results and also your solution if it might be interesting for the others.

**Comment by frangarcj** [comment]
Would you mind adding the hacktoberfest-accepted label to the PR? It's for getting some swag.

Thanks in advance.

**Comment by mkurzydlowski** [comment]
Here you are.

**Comment by ctu-developers** [comment]
@frangarcj great job! It also bothered us and we wanted to solve it. Thanks.

**Comment by Emkas** [comment]
Really nicely done! Time reduction is great.

**Pull request merged**